source/priority_queue.c

```diff
@@ -0,0 +1,139 @@
+
+    while(left < len) {
+        aws_array_list_get_at_ptr(&queue->container, &left_item, left);
+
+        size_t right = left + 1;
+        if(right < len) {
+            aws_array_list_get_at_ptr(&queue->container, &right_item, right);
+
+            /* choose the larger/smaller of the two in case of a max/min heap respectively */
+            if (queue->compare(left_item, right_item)) {
+                left = right;
+                left_item = right_item;
+            }
+        }
+
+        aws_array_list_get_at_ptr(&queue->container, &root_item, root);
+        if(queue->compare(root_item, left_item)) {
+            aws_array_list_swap(&queue->container, left, root);
+            root = left;
+            left = left_of(root);
+        }
+        else {
+            break;
+        }
+    }
+}
+
+/* Precondition: Elements prior to the specified index must be in heap order. */
+static void sift_up(struct aws_priority_queue *queue, size_t index) {
+    void *parent_item, *child_item;
+    size_t parent = parent_of(index);
+    while(index) {
+        aws_array_list_get_at_ptr(&queue->container, &parent_item, parent);
+        aws_array_list_get_at_ptr(&queue->container, &child_item, index);
+
+        if (queue->compare(parent_item, child_item)) {
+            aws_array_list_swap(&queue->container, index, parent);
+            index = parent;
+            parent = parent_of(index);
+        }
+        else {
+            break;
+        }
+    }
+}
+
+int aws_priority_queue_dynamic_init(struct aws_priority_queue *queue, struct aws_allocator *alloc, size_t default_size,
+                                    size_t item_size, int(*compare)(void *a, void *b)) {
+
+    queue->compare = compare;
+    return aws_array_list_init_dynamic(&queue->container, alloc, default_size, item_size);
+}
+
+void aws_priority_queue_static_init(struct aws_priority_queue *queue,
+    void *heap, size_t item_count, size_t item_size, int(*compare)(void *a, void *b)) {
+
+    queue->compare = compare;
+    aws_array_list_init_static(&queue->container, heap, item_count, item_size);
+}
+
+void aws_priority_queue_clean_up(struct aws_priority_queue *queue) {
+    aws_array_list_clean_up(&queue->container);
+}
+
+int aws_priority_queue_push(struct aws_priority_queue *queue, void* item) {
+    int err = aws_array_list_push_back(&queue->container, item);
+    if(err) {
+        return err;
+    }
+
+    sift_up(queue, aws_array_list_length(&queue->container) - 1);
+
+    return AWS_OP_SUCCESS;
+}
+
+int aws_priority_queue_pop(struct aws_priority_queue *queue, void* item) {
+
+    if(0 == aws_array_list_length(&queue->container)) {
+        return AWS_ERROR_PRIORITY_QUEUE_EMPTY;
+    }
+
+    int err = aws_priority_queue_top(queue, item);
+
+    if (err) {
+        return err;
+    }
+
+    aws_array_list_swap(&queue->container, 0, aws_array_list_length(&queue->container) - 1);
+
+    err = aws_array_list_pop_back(&queue->container);
```

Comment:
nit:

/* because in this case aws_raise_error has already been called */
if (aws_array_list_*(&queue->container)) {
    return AWS_OP_ERR;
}

Reply:
sure